src/character.cpp

```diff
@@ -9684,6 +9684,44 @@ bool Character::use_charges_if_avail( const itype_id &it, int quantity )
     return false;
 }
 
+std::pair<float, item> Character::get_best_weapon_by_damage_type( const damage_type_id dmg_type )
+const
+{
+    std::pair<float, item> best_weapon = std::make_pair( 0, item() );
+
+    cache_visit_items_with( "best_damage", &item::is_melee, [&best_weapon, this,
+                  dmg_type]( const item & it ) {
+        damage_instance di;
+        roll_damage( dmg_type, false, di, true, it, attack_vector_id::NULL_ID(),
+                     sub_bodypart_str_id::NULL_ID(), 1.f );
+        for( damage_unit &du : di.damage_units ) {
+            if( du.type == dmg_type && best_weapon.first < du.amount ) {
+                best_weapon = std::make_pair( du.amount, it );
+            }
+        }
+    } );
+
+    return best_weapon;
+}
+
+std::pair<int, item> Character::get_best_tool( const quality_id quality ) const
```

Comment:
I think we want to return `const item&` (or `const item *`) to avoid neeless copies. It's not in the hot path currently, so not a huge deal, but `item` is a chunky struct still.